paddle/fluid/pir/dialect/operator/ir/ops.yaml

```
@@ -387,6 +387,15 @@
     data_type : fpn_rois
   optional : rois_num, multi_level_rois_num
 
+- op : distributed_fused_lamb_init
+  args : (Tensor[] param, Tensor[] grad, float beta1, float beta2, int[] apply_weight_decay, int alignment, int rank, int nranks)
+  output : Tensor(fp32_fused_param), Tensor(fp32_fused_grad), Tensor(fp16_fused_param), Tensor(fp16_fused_grad), Tensor(moment1), Tensor(moment2), Tensor(beta1_pow), Tensor(beta2_pow), Tensor(fused_param_offsets), Tensor(fp32_shard_fused_param_offsets), Tensor(fp16_shard_fused_param_offsets), Tensor(param_info), Tensor(param_order), Tensor[](param_out){param.size()}, Tensor[](master_param_out){param.size()}, Tensor[](grad_out){grad.size()}, Tensor(global_scale), Tensor(step)
+  infer_meta :
+    func : DistributedFusedLambInitInferMeta
+  kernel :
+    func : distributed_fused_lamb_init
+  optional : fp32_fused_param, fp32_fused_grad, fp16_fused_param, fp16_fused_grad
```

如果某个变量同时在输入输出中出现，意味着它是inplace的，这里主要是原Op注册时没有正确声明。对于这样的参数，可以像这样补充下设置：
```suggestion
  optional : fp32_fused_param, fp32_fused_grad, fp16_fused_param, fp16_fused_grad
  inplace:  (param -> param_out, grad -> grad_out) 
```

---

请问一下这里怎么判断是不是inplace呢，我看了一下只能通过编译才知道会有‘_’的需求，在原有文件上似乎没看到有这种提示

---

@kangguangli 
另外加上这个inplace后，报错了找不到eager_**相关的函数，我看python_gen没太明白这里为什么跳过了eager_*函数的生成，static函数是正常生成的，是ops.yaml里面相关参数还是填错了吗？
```
/home/aistudio/http/Paddle/paddle/fluid/pybind/ops_api.cc:3825:12: error: ‘eager_api_distributed_fused_lamb_init’ was not declared in this scope
     return eager_api_distributed_fused_lamb_init(self, args, kwargs);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/aistudio/http/Paddle/paddle/fluid/pybind/ops_api.cc:3825:12: note: suggested alternative: ‘static_api_distributed_fused_lamb_init’
     return eager_api_distributed_fused_lamb_init(self, args, kwargs);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            static_api_distributed_fused_lamb_init
/home/aistudio/http/Paddle/paddle/fluid/pybind/ops_api.cc: In function ‘PyObject* paddle::pybind::distributed_fused_lamb_init_(PyObject*, PyObject*, PyObject*)’:
/home/aistudio/http/Paddle/paddle/fluid/pybind/ops_api.cc:3834:12: error: ‘eager_api_distributed_fused_lamb_init_’ was not declared in this scope
     return eager_api_distributed_fused_lamb_init_(self, args, kwargs);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/aistudio/http/Paddle/paddle/fluid/pybind/ops_api.cc:3834:12: note: suggested alternative: ‘static_api_distributed_fused_lamb_init_’
     return eager_api_distributed_fused_lamb_init_(self, args, kwargs);
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            static_api_distributed_fused_lamb_init_
```

---

> 请问一下这里怎么判断是不是inplace呢，我看了一下只能通过编译才知道会有‘_’的需求，在原有文件上似乎没看到有这种提示

这里跟append_op时的做法有关，inputs/outputs中使用了同样的变量params/grads，就说明我们需要这个Op是inplace的。如果输出中使用其他变量，就不需要声明。这里还是需要声明一下。

> 没太明白这里为什么跳过了eager_*函数的生成

这里ops_api.cc是被`paddle/fluid/pir/dialect/op_generator/ops_api_gen.py`控制生成的，它默认Op同时有静态图版本和动态图版本。静态图API是`paddle/fluid/pir/dialect/op_generator/python_c_gen.py`生成，动态图API是`paddle/fluid/eager/auto_code_generator/generator/python_c_gen.py`，它们是不一致的。

解决办法是在`paddle/fluid/pir/dialect/op_generator/ops_api_gen.py`的NEED_GEN_STATIC_ONLY_APIS中添加 distributed_fused_lamb_init和distributed_fused_lamb_init_。

当然这里有个问题是为什么加了inplace以后才报错，目前我也没想通，可以先试试这样能否解决。

---

多谢，我有空改一下